source/structuring.rst

```diff
@@ -14,402 +14,545 @@ functions is better than copying-and-pasting duplicate bits of code
+      "$id": "https://example.com/schemas/customer",
+
+      "type": "object",
+      "properties": {
+        "first_name": { "type": "string" },
+        "last_name": { "type": "string" },
+        "shipping_address": { "$ref": "/schemas/address" },
+        "billing_address": { "$ref": "/schemas/address" }
+      },
+      "required": ["first_name", "last_name", "shipping_address", "billing_address"],
 
       "definitions": {
         "address": {
+          "$id": "/schemas/address",
+
           "type": "object",
           "properties": {
             "street_address": { "type": "string" },
-            "city":           { "type": "string" },
-            "state":          { "type": "string" }
+            "city": { "type": "string" },
+            "state": { "$ref": "#/definitions/state" }
           },
-          "required": ["street_address", "city", "state"]
-        }
-      },
-
-      "type": "object",
+          "required": ["street_address", "city", "state"],
 
-      "properties": {
-        "billing_address": { "$ref": "#/definitions/address" },
-        "shipping_address": {
-          "allOf": [
-            { "$ref": "#/definitions/address" },
-            { "properties":
-              { "type": { "enum": [ "residential", "business" ] } },
-              "required": ["type"]
-            }
-          ]
+          "definitions": {
+            "state": { "enum": ["CA", "NY", "... etc ..."] }
+          }
         }
       }
     }
-    --X
-    // This fails, because it's missing an address type:
-    {
-      "shipping_address": {
-        "street_address": "1600 Pennsylvania Avenue NW",
-        "city": "Washington",
-        "state": "DC"
-      }
-    }
-    --
-    {
-      "shipping_address": {
-        "street_address": "1600 Pennsylvania Avenue NW",
-        "city": "Washington",
-        "state": "DC",
-        "type": "business"
-      }
-    }
 
-From these basic pieces, it's possible to build very powerful
-constructions without a lot of duplication.
+Notice that the ``$ref`` keywords from the customer schema resolve the
+same way they did before except that the address schema is now defined
+at ``/definitions/address`` instead of a separate schema document. You
+should also see that the ``"$ref": "#/definitions/state"`` resolves to
+the ``definitions`` keyword in the address schema rather than the one
+at the top level schema like it would if the subschema ``$id`` wasn't
+used.
+
+You might notice that this creates a situation where there are
+multiple ways to identify a schema. Instead of referencing
+``/schemas/address`` (``https://example.com/schemas/address``) I could
+have used ``#/definitions/address``
+(``https://example.com/schemas/customer#/definitions/address``). While
+both of these will work, the one shown in the example is preferred
+because the  ``$id`` keyword establishes
+``https://example.com/schemas/address`` as the canonical URI for the
+schema being referenced.
```

Comment:
Only the canonical URI is guaranteed to work.

Implementations MAY provide support for non-canonical URI references, but it is not required.

Reply:
Remember this is for draft-07. Allowing implementations to not support non-canonical URIs was added in 2019-09.

Reply:
We didn't define what a canonical URI meant for JSON Schema in draft-07.
But I understand "canonical URI" still has a meaning.